audio-worklet/src/noisegate-script-processor.js

```diff
@@ -0,0 +1,179 @@
+/**
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+class NoiseGate {
+  /**
+   * A Noise gate allows audio signals to pass only when the registered volume
+   * is above a specified threshold.
+   * @param  {BaseAudioContext} context the audio context
+   * @param  {Number} options.channels the number of input and output 
+   *                                   channels, a maximum of two
+   * @param  {Object} options parameters for the noise gate                                  
+   * @param  {Number} options.attack seconds for gate to fully close  
+   * @param  {Number} options.release seconds for gate to fully open
+   * @param  {Number} options.bufferSize the size of an onaudioprocess window
+   * @param  {Number} options.threshold gain level beneath which sound is muted
+   * @param  {Number} options.alpha weight for the envelop follower affecting
+   *                                the tradeoff between delay and smoothness
+   */
+  constructor(context, options) {
+    if (!(context instanceof BaseAudioContext)) 
+      throw context + ' is not a valid audio context.';
+    if (!options) options = {};
+    const numberOfChannels = options.numberOfChannels || 1; 
+    if (numberOfChannels > 2) 
+      throw 'The maximum supported number of channels is two.';
+
+    this.context_ = context;
+    const bufferSize = options.bufferSize || 0;
+    this.threshold = options.threshold || 1;
+    this.attack = options.attack || 0;
+    this.release = options.release || 0;
+
+    // Alpha controls a tradeoff between the smoothness of the 
+    // envelope and its delay, with a higher value giving more smoothness at 
+    // the expense of delay and vice versa. It has been set experimentally to
+    // control this tradeoff.
+    this.alpha_ = options.alpha || 0.90;
+
+    // Alpha is normalized around the minimum sample rate of an offline audio
+    // context (3000). 
+    this.alpha_ = 1 -
+        ((1 - this.alpha_) / (this.context_.sampleRate / 3000));
```

Comment:
You're going to have to prove that this is correct. I'm not smart enough to understand what this is doing and why it should be this way.